File: ios-renderer/Sources/A2UIRenderer/Processing/TokenResolver.swift
import Foundation
import SwiftUI

/// Resolves design token references like "{Accents.Red}" → "#ff4245".
/// Port of resolveToken() from a2ui-parser/src/index.ts.
public struct TokenResolver {
    public let tokens: [String: DesignToken]

    public init(tokens: [String: DesignToken]) {
        self.tokens = tokens
    }

    /// Resolve a value that may be a token reference.
    /// - "{Accents.Red}" → token's value ("#ff4245")
    /// - "#ff4245" → "#ff4245" (passthrough)
    /// - 16 → 16 (passthrough)
    public func resolve(_ value: JSONValue?) -> JSONValue? {
        guard let value else { return nil }
        guard let str = value.stringValue else { return value }

        // Check for token reference pattern: {token.name}
        guard str.hasPrefix("{"), str.hasSuffix("}"),
              !str.hasPrefix("${") else { return value }

        let tokenName = String(str.dropFirst().dropLast())
        if let token = tokens[tokenName] {
            // Try to return as number if it's a numeric value
            if let num = Double(token.value) {
                return .number(num)
            }
            return .string(token.value)
        }
        return value
    }

    /// Resolve a style property value.
    public func resolveStyle(_ style: [String: JSONValue]?, key: String) -> JSONValue? {
        guard let val = style?[key] else { return nil }
        return resolve(val)
    }

    /// Resolve a style property as CGFloat.
    public func resolveStyleCGFloat(_ style: [String: JSONValue]?, key: String) -> CGFloat? {
        resolveStyle(style, key: key)?.asCGFloat
    }

    /// Resolve a style property as String.
    public func resolveStyleString(_ style: [String: JSONValue]?, key: String) -> String? {
        resolveStyle(style, key: key)?.asStyleValue
    }

    /// Resolve a color from a style property (hex string → Color).
    public func resolveColor(_ style: [String: JSONValue]?, key: String) -> Color? {
        guard let hex = resolveStyleString(style, key: key) else { return nil }
        return Color(hex: hex)
    }
}

// MARK: - Color from hex

extension Color {
    public init?(hex: String) {
        var hexString = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        if hexString.hasPrefix("#") {
            hexString.removeFirst()
        }

        guard hexString.count == 6 || hexString.count == 8 else { return nil }

        var rgb: UInt64 = 0
        Scanner(string: hexString).scanHexInt64(&rgb)

        if hexString.count == 8 {
            self.init(
                red: Double((rgb >> 24) & 0xFF) / 255.0,
                green: Double((rgb >> 16) & 0xFF) / 255.0,
                blue: Double((rgb >> 8) & 0xFF) / 255.0,
                opacity: Double(rgb & 0xFF) / 255.0
            )
        } else {
            self.init(
                red: Double((rgb >> 16) & 0xFF) / 255.0,
                green: Double((rgb >> 8) & 0xFF) / 255.0,
                blue: Double(rgb & 0xFF) / 255.0
            )
        }
    }
}

File: android-renderer/lib/src/main/kotlin/com/interfaceai/a2ui/processing/TokenResolver.kt
package com.interfaceai.a2ui.processing

import androidx.compose.ui.graphics.Color
import com.interfaceai.a2ui.models.DesignToken
import kotlinx.serialization.json.*

/**
 * Resolves design token references like "{Accents.Red}" → "#ff4245".
 * Port of resolveToken() from a2ui-parser/src/index.ts.
 */
class TokenResolver(val tokens: Map<String, DesignToken>) {

    /**
     * Resolve a JSON value that may be a token reference.
     */
    fun resolve(value: JsonElement?): JsonElement? {
        if (value == null) return null
        val str = (value as? JsonPrimitive)?.contentOrNull ?: return value

        // Check for token reference: {token.name}
        if (!str.startsWith("{") || !str.endsWith("}") || str.startsWith("\${")) return value

        val tokenName = str.substring(1, str.length - 1)
        val token = tokens[tokenName] ?: return value

        // Return as number if numeric, else as string
        token.value.toDoubleOrNull()?.let {
            return JsonPrimitive(it)
        }
        return JsonPrimitive(token.value)
    }

    /**
     * Resolve a style property value.
     */
    fun resolveStyle(style: Map<String, JsonElement>?, key: String): JsonElement? {
        val value = style?.get(key) ?: return null
        return resolve(value)
    }

    /**
     * Resolve a style property as Float.
     */
    fun resolveStyleFloat(style: Map<String, JsonElement>?, key: String): Float? {
        val resolved = resolveStyle(style, key) ?: return null
        return when (resolved) {
            is JsonPrimitive -> resolved.floatOrNull ?: resolved.contentOrNull?.toFloatOrNull()
            else -> null
        }
    }

    /**
     * Resolve a style property as String.
     */
    fun resolveStyleString(style: Map<String, JsonElement>?, key: String): String? {
        val resolved = resolveStyle(style, key) ?: return null
        return (resolved as? JsonPrimitive)?.contentOrNull
    }

    /**
     * Resolve a color from a hex string style property.
     */
    fun resolveColor(style: Map<String, JsonElement>?, key: String): Color? {
        val hex = resolveStyleString(style, key) ?: return null
        return parseHexColor(hex)
    }
}

/**
 * Parse a hex color string like "#ff4245" into a Compose Color.
 */
fun parseHexColor(hex: String): Color? {
    val clean = hex.removePrefix("#")
    return try {
        when (clean.length) {
            6 -> Color(android.graphics.Color.parseColor("#$clean"))
            8 -> Color(android.graphics.Color.parseColor("#$clean"))
            else -> null
        }
    } catch (e: Exception) {
        null
    }
}
